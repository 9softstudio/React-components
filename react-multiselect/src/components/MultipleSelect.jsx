import React, { Component } from 'react';
import PropTypes from 'prop-types';

import MultipleSelectLabel from './MultipleSelectLabel'
import MultipleSelectOptionList from './MultipleSelectOptionList'
import OptionAll from './OptionAll'

const KEY_NAME = "key";
const VALUE_NAME = "value";
const STATUS_NAME = "checked";

let id = 1;

export default class MultipleSelect extends Component {
    constructor(props) {
        super(props);
        this.searchInputBox = null;
        this.id = `multiple-select-${props.id || id++}`;
        this.originalDataSource = this._convertDataSourceToState(props);

        this.state = {
            searchText: '',
            showOptionList: false,
            dataSource: this.originalDataSource
        }
    }

    static propTypes = {
        dataSource: PropTypes.arrayOf(PropTypes.object).isRequired,
        keyField: PropTypes.string,
        valueField: PropTypes.string,
        statusField: PropTypes.string,
        noneSelectedLabel: PropTypes.string,
        optionAllLabel: PropTypes.string,
        maxDisplayItemCount: PropTypes.number,
        onChange: PropTypes.func,
        hasAllOption: PropTypes.bool,
        hasSearchBox: PropTypes.bool
    }

    static defaultProps = {
        keyField: KEY_NAME,
        valueField: VALUE_NAME,
        statusField: STATUS_NAME,
        hasAllOption: true,
        hasSearchBox: false
    }

    componentDidMount() {
        document.addEventListener('click', this._handleDocumentClick)
    }

    componentWillUnmount() {
        document.removeEventListener('click', this._handleDocumentClick)
    }

    get selectedItems() {
        return this.originalDataSource.filter(item => item.checked);
    }

    onChangeHandler = (item) => {
        const { keyField, valueField, statusField } = this.props;

        const selectedItem = {
            [keyField]: item.key,
            [valueField]: item.value,
            [statusField]: item.checked
        }

        const newDataSource = this.state.dataSource.slice(0);
        const itemToUpdate = newDataSource.find(x => x.key === item.key);
        itemToUpdate.checked = item.checked;

        this._callBackToParent(selectedItem);

        this.setState({ dataSource: newDataSource });
    }
    
    onToggle = () => {
        this.setState((prevState) => { return { showOptionList: !prevState.showOptionList } },
            () => {
                if(this.searchInputBox && this.state.showOptionList){
                    this.searchInputBox.focus();
                }
            })
    }

    checkAllHandler = (checked) => {
        this.originalDataSource.forEach((item) => { item.checked = checked });

        const searchText = this.state.searchText;
        const dataSource = this.originalDataSource.filter(item => item[VALUE_NAME].startsWith(searchText));
        this.setState({ dataSource: dataSource });
        this._callBackToParent(null);
    }

    onChangeSearchText = (value) => {
        this.setState((prevState) => {
            const dataSource = this.originalDataSource.filter(item => item[VALUE_NAME].startsWith(value));

            return { 
                searchText: value,
                dataSource : dataSource
            }
        })
    }

    onClearSearch = () => {
        this.setState({searchText: ''}, () => this.onChangeSearchText(''))
    }

    _callBackToParent(selectedItem) {
        const selectedItemsKey = this._getSelectedItemKey(this.originalDataSource);
        this.props.onChange && this.props.onChange(selectedItem, selectedItemsKey);
    }

    _convertDataSourceToState({ keyField, valueField, statusField, dataSource }) {
        return dataSource.map(item => {
            return {
                [KEY_NAME]: item[keyField],
                [VALUE_NAME]: item[valueField],
                [STATUS_NAME]: item[statusField]
            }
        })
    }

    _getSelectedItemKey(dataSource) {
        const selectedItemKey = [];

        for (let i = 0; i < dataSource.length; i++) {
            const item = dataSource[i];

            if (item.checked) {
                selectedItemKey.push(item.key)
            }
        }

        return selectedItemKey.join(",");
    }

    _close = () => {
        this.setState({ showOptionList: false });
    }

    _handleDocumentClick = (event) => {
        const clickOutside = this.wrapper && !this.wrapper.contains(event.target);
        if (clickOutside && this.state.showOptionList) {
            this._close();
        }
    }

    _renderOptionAll() {
        if (this.props.hasAllOption) {
            const checkedItemCount = this.originalDataSource.filter(x => x.checked).length;
            const checkedAll = checkedItemCount === this.originalDataSource.length;
            
            return (<OptionAll id={this.id} checked={checkedAll} label={this.props.optionAllLabel} onChange={this.checkAllHandler} />);
        }

        return null;
    }

    _renderSearchBox(){
        if(!this.props.hasSearchBox)
        {
            return null;
        }

        const clearButtonStyle =  { 
            position: "absolute",
            right: "5px",
            width: "20px",
            height: "20px",
            border: 0,
            margin: "5px",
            backgroundColor: "transparent",
            cursor: "pointer"}

        return (<div style={{width:"100%"}}>
                        <input autoFocus  onChange={(e) => this.onChangeSearchText(e.target.value)} 
                            ref={element => this.searchInputBox = element}
                            value={this.state.searchText} type="text" placeholder="Search data" 
                            style={{width:"100%", border: "1px solid #ccc", padding: "5px 10px", 
                            boxSizing: "border-box",}} />
        <button onClick={this.onClearSearch} style={clearButtonStyle}>X</button>
        </div>)
    }

    render() {
        const { noneSelectedLabel, maxDisplayItemCount } = this.props;
    
        return (
            <div className="multiple-select-container" id={this.id} ref={element => this.wrapper = element}>
                 {
                <MultipleSelectLabel
                    selectedItems={this.selectedItems}
                    onToggle={this.onToggle}
                    noneSelectedLabel={noneSelectedLabel}
                    maxDisplayItemCount={maxDisplayItemCount} />
                 }
                <div className="multiple-select-default multiple-select-options-container"
                style={{ display: this.state.showOptionList ? "block" : "none" }}>
                    {this._renderSearchBox()}
                        {this._renderOptionAll()}
                    <MultipleSelectOptionList id={this.id}
                        dataSource={this.state.dataSource}
                        onChange={this.onChangeHandler} />
                </div>
            </div>
        )
    }
}