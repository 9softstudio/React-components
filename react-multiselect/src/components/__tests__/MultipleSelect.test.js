import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15'

import MultipleSelect from '../MultipleSelect'

configure({ adapter: new Adapter() });

const defaultProps = {
    dataSource: [],
    keyField: "key",
    valueField: "value",
    statusField: "checked",
    noneSelectedLabel: "Select Options",
    optionAllLabel: "All",
    maxDisplayItemCount: 3,
    onChange: jest.fn(),
    hasAllOption: true
}

const defaultOption = { code: "SB", name: "Sportsbook", status: true };
const defaultDataSource = [defaultOption];

describe('MultipleSelect ', () => {
    afterAll(() => {
        jest.resetAllMocks();
    })

    describe('Initial State', () => {
        it('has id prop - id value should be correct', () => {
            let defaultId = 1;

            const testIdProp = (idValue) => {
                const currentId = idValue || defaultId++;
                const expectedResult = `multiple-select-${currentId}`;
                const multipleSelect = renderer.create(<MultipleSelect id={idValue} {...defaultProps} />);
                const instance = multipleSelect.getInstance();

                const result = instance.id;

                expect(result).toBe(expectedResult);
            }

            const idValues = ["multipleSelectId", undefined, undefined, 100];

            idValues.forEach(testIdProp);
        })

        it('showOptionList should be false', () => {
            const multipleSelect = new MultipleSelect(defaultProps);

            const result = multipleSelect.state.showOptionList;

            expect(result).toBeFalsy();
        })

        it('dataSource prop should convert correctly', () => {
            const expectedResult = [
                { key: defaultOption.code, value: defaultOption.name, checked: defaultOption.status }
            ]

            const newProp = {
                dataSource: defaultDataSource,
                keyField: "code",
                valueField: "name",
                statusField: "status"
            }

            const props = Object.assign({}, defaultProps, newProp);
            const multipleSelect = new MultipleSelect(props);

            const result = multipleSelect.state.dataSource;

            expect(result).toEqual(expectedResult);
        })
    })

    describe('selectedItems', () => {
        it('selectedItems should return correct data', () => {
            const sportsbookOption = { key: "SB", value: "Sportsbook", checked: true };
            const expectedResult = [sportsbookOption];
            const dataSource = [sportsbookOption, { key: "BA", value: "BA", checked: false }];

            const props = Object.assign({}, defaultProps, { dataSource });
            const multipleSelect = new MultipleSelect(props);

            const result = multipleSelect.selectedItems;

            expect(result).toEqual(expectedResult);
        })
    })

    describe('onChangeHandler', () => {
        let multipleSelect;

        const sportsbookOption = { key: "SB", value: "Sportsbook", checked: true };
        const changedItem = { key: "BA", value: "BA", checked: true };

        const dataSource = [sportsbookOption, Object.assign({}, changedItem, { checked: false })];
        const newDataSource = [sportsbookOption, changedItem];

        beforeEach(() => {
            const props = Object.assign({}, defaultProps, { dataSource });
            multipleSelect = new MultipleSelect(props);

            multipleSelect._callBackToParent = jest.fn();
            multipleSelect.setState = jest.fn();

            multipleSelect.onChangeHandler(changedItem);
        })

        it('_callBackToParent should be called with correct parameter', () => {
            expect(multipleSelect._callBackToParent).toHaveBeenCalledWith(changedItem);
        })

        it('setState should be called with correct parameter', () => {
            const newState = { dataSource: newDataSource };

            expect(multipleSelect.setState).toHaveBeenCalledWith(newState);
        })
    })

    describe('onToggle', () => {
        it('new State should be correct', () => {
            let callback;
            const multipleSelect = new MultipleSelect(defaultProps);
            multipleSelect.setState = jest.fn((cb) => callback = cb);

            multipleSelect.onToggle();

            let result = callback({ showOptionList: false });
            expect(result).toEqual({ showOptionList: true });

            result = callback({ showOptionList: true });
            expect(result).toEqual({ showOptionList: false });
        })
    })

    describe('checkAllHandler', () => {
        let multipleSelect;
        const sportsbookOption = { key: "SB", value: "Sportsbook", checked: true };
        const baOption = { key: "BA", value: "BA", checked: true };

        const testCheckAll = (checked) => {
            const dataSource = [
                Object.assign({}, sportsbookOption, { checked: !checked }),
                Object.assign({}, baOption, { checked: !checked })
            ];

            const props = Object.assign({}, defaultProps, { dataSource });
            multipleSelect = new MultipleSelect(props);

            multipleSelect._callBackToParent = jest.fn();
            multipleSelect.setState = jest.fn();

            multipleSelect.checkAllHandler(checked);
        }

        it('_callBackToParent should be called with correct parameter', () => {
            const checked = true;
            const newDataSource = [
                Object.assign({}, sportsbookOption, { checked }),
                Object.assign({}, baOption, { checked })
            ];

            testCheckAll(checked);

            expect(multipleSelect._callBackToParent).toHaveBeenCalledWith(null);
        })

        it('uncheck All - setState should be called with correct parameter', () => {
            const checked = false;
            const newDataSource = [
                Object.assign({}, sportsbookOption, { checked }),
                Object.assign({}, baOption, { checked })
            ];

            const newState = { dataSource: newDataSource };

            testCheckAll(checked);

            expect(multipleSelect.setState).toHaveBeenCalledWith(newState);
        })

        it('check All - setState should be called with correct parameter', () => {
            const checked = true;
            const newDataSource = [
                Object.assign({}, sportsbookOption, { checked }),
                Object.assign({}, baOption, { checked })
            ];

            const newState = { dataSource: newDataSource };

            testCheckAll(checked);

            expect(multipleSelect.setState).toHaveBeenCalledWith(newState);
        })
    })

    describe('onChangeSearchText', () => {
        let element;
        let setStateCallback;
        beforeEach(() => {
            element = new MultipleSelect(defaultProps);
            element.originalDataSource = [{value: "test123"}];
            
            element.setState = jest.fn((callback) => {
                setStateCallback = callback;
            });
        });

        it('input value does not match any value in data source: setState() with correct param', () => {
            element.onChangeSearchText("xxx");
            const result = setStateCallback();
            
            expect(result).toEqual({"dataSource": [], "searchText": "xxx"});
        })

        it('input value matches values in data source: setState() with correct param', () => {
            element.onChangeSearchText("test");
            const result = setStateCallback();
            
            expect(result).toEqual({"dataSource": [{"value": "test123"}], "searchText": "test"});
        })
    })

    describe('onClearSearch', () => {
        it('call setState() and onChangeSearchText() with correct parameters', () => {
            const multipleSelect = new MultipleSelect(defaultProps);
            let setStateData;
            let setStateCallback;
            multipleSelect.setState = jest.fn((data, callback) => {
                setStateData = data;
                setStateCallback = callback;
            });
            multipleSelect.onChangeSearchText = jest.fn();

            multipleSelect.onClearSearch();
            setStateCallback();

            expect(setStateData).toEqual({searchText: ''});
            expect(multipleSelect.onChangeSearchText).toHaveBeenCalledWith('');
        })
    })

    describe('_callBackToParent', () => {
        it('call onChange function in props with correct parameters', () => {
            const expectedSelectedItemKey = "SB,BA";
            const expectedSelectedItem = { key: "Key", value: "Value", checked: true };
            const multipleSelect = new MultipleSelect(defaultProps);

            multipleSelect._getSelectedItemKey = jest.fn().mockReturnValueOnce(expectedSelectedItemKey);
            multipleSelect.props.onChange = jest.fn();

            const dataSource = [];
            multipleSelect._callBackToParent(expectedSelectedItem, dataSource);

            expect(multipleSelect.props.onChange).toHaveBeenCalledWith(expectedSelectedItem, expectedSelectedItemKey);
        })
    })

    describe('_getSelectedItemKey', () => {
        it('get correct key of items that checked', () => {
            const expectedResult = "SB,BA";
            const dataSource = [
                { key: "SB", value: "Value SB", checked: true },
                { key: "BA", value: "Value BA", checked: true },
                { key: "RC", value: "Value RC", checked: false }
            ]

            const componentInstance = new MultipleSelect(defaultProps);

            const result = componentInstance._getSelectedItemKey(dataSource);

            expect(result).toBe(expectedResult);
        })
    })

    describe('close', () => {
        it('setState function should be called with correct parameter', () => {
            const expectedParam = { showOptionList: false };

            const componentInstance = new MultipleSelect(defaultProps);
            componentInstance.setState = jest.fn();

            componentInstance._close();

            expect(componentInstance.setState).toHaveBeenCalledWith(expectedParam);
        })
    })

    describe('_handleDocumentClick', () => {
        it('wrapper element is null - do nothing', () => {
            const instance = new MultipleSelect(defaultProps);
            instance._close = jest.fn();

            const clickEvent = new Event('click');
            document.dispatchEvent(clickEvent);

            expect(instance._close).not.toHaveBeenCalled();
        })

        it('showOptionList is false - do nothing', () => {
            const component = mount(<MultipleSelect {...defaultProps} />);
            const instance = component.instance();

            instance._close = jest.fn();

            const clickEvent = new Event('click');
            document.dispatchEvent(clickEvent);

            expect(instance._close).not.toHaveBeenCalled();
        })

        it('showOptionList is true - call _close function', () => {
            const component = mount(<MultipleSelect {...defaultProps} />);
            component.setState({ showOptionList: true });
            const instance = component.instance();

            instance._close = jest.fn();

            const clickEvent = new Event('click');
            document.dispatchEvent(clickEvent);

            expect(instance._close).toHaveBeenCalled();
        })
    })

    describe('_renderOptionAll', () => {
        it('hassAllOption in props is false - return null', () => {
            const newProps = Object.assign({}, defaultProps, { hasAllOption: false });
            const componentInstance = new MultipleSelect(newProps);

            const result = componentInstance._renderOptionAll();

            expect(result).toBe(null);
        })

        it('hasAllOption in props is true - render OptionAll component with correct props', () => {
            const dataSource = [
                { key: "SB", value: "Value SB", checked: true },
                { key: "BA", value: "Value BA", checked: true },
                { key: "RC", value: "Value RC", checked: true }
            ];
            const newProps = Object.assign({}, defaultProps, { dataSource });
            const componentInstance = new MultipleSelect(newProps);
            const optionAllComponent = componentInstance._renderOptionAll();

            const { checked } = optionAllComponent.props;

            expect(checked).toBe(true);
        })
    })

    describe('render', () => {
        it('showOptionList = false - render component with "display: none" for dropdown list', () => {
            const component = renderer.create(<MultipleSelect {...defaultProps} />);

            const tree = component.toJSON();

            expect(tree).toMatchSnapshot();
        })

        it('showOptionList = true - render component with "display: block" for dropdown list', () => {
            const component = renderer.create(<MultipleSelect {...defaultProps} />);
            const instance = component.getInstance();
            instance.setState({ showOptionList: true });

            const tree = component.toJSON();

            expect(tree).toMatchSnapshot();
        })
    })

    describe('componentWillUnmount', () => {
        it('document.removeEventListener function should be called with correct event name', () => {
            const originalRemoveEventListenerFunction = document.removeEventListener;
            const componentInstance = new MultipleSelect(defaultProps);
            document.removeEventListener = jest.fn();

            componentInstance.componentWillUnmount();

            expect(document.removeEventListener.mock.calls[0][0]).toBe('click');

            document.removeEventListener = originalRemoveEventListenerFunction;
        })
    })

    describe('componentWillReceiveProps', () => {
        it('dataSource is the same - do nothing', () => {
            const nextProps = Object.assign({}, defaultProps);
            const multipleSelect = new MultipleSelect(defaultProps);
            multipleSelect.setState = jest.fn();

            multipleSelect.componentWillReceiveProps(nextProps);

            expect(multipleSelect.setState).not.toHaveBeenCalled();
        })

        it('dataSource is updated - setState should be called with correct parameter', () => {
            const props = Object.assign({}, defaultProps, { dataSource: defaultDataSource });
            const nextProps = Object.assign({}, props);
            const multipleSelect = new MultipleSelect(defaultProps);
            multipleSelect.setState = jest.fn();
            multipleSelect._convertDataSourceToState = jest.fn(() => defaultDataSource);
            multipleSelect._callBackToParent = jest.fn();

            const expectedState = {
                dataSource: defaultDataSource
            }

            multipleSelect.componentWillReceiveProps(nextProps);

            expect(multipleSelect.setState).toHaveBeenCalledWith(expectedState);
        })

        it('dataSource is updated - _callBackToParent should be called with correct parameter', () => {
            const props = Object.assign({}, defaultProps, { dataSource: defaultDataSource });
            const nextProps = Object.assign({}, props);
            const multipleSelect = new MultipleSelect(defaultProps);
            multipleSelect.setState = jest.fn();
            multipleSelect._convertDataSourceToState = jest.fn(() => defaultDataSource);
            multipleSelect._callBackToParent = jest.fn();

            multipleSelect.componentWillReceiveProps(nextProps);

            expect(multipleSelect._callBackToParent).toHaveBeenCalledWith(null);
        })
    })
})