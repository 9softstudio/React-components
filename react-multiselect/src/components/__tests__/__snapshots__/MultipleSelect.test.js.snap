// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleSelect  render showOptionList = false - render component with "display: none" for dropdown list 1`] = `
<div
  className="multiple-select-container"
  id="multiple-select-23"
>
  <button
    className="multiple-select-default multiple-select-label"
    onClick={[Function]}
    type="button"
  >
    <span
      className="text-display"
      title="Select Options"
    >
      Select Options
    </span>
    <b
      className="caret"
    />
  </button>
  <div
    className="multiple-select-default multiple-select-options-container"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <ul
      className="multiple-select-options multiple-select-option-all"
    >
      <li
        className="multiple-select-item"
      >
        <input
          checked={true}
          className="option-checkbox"
          id="multiple-select-23-optionItemAll"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="option-label"
          htmlFor="multiple-select-23-optionItemAll"
        >
          (All)
        </label>
      </li>
    </ul>
    <ul
      className="multiple-select-options"
    >
      <span>
        No information
      </span>
    </ul>
  </div>
</div>
`;

exports[`MultipleSelect  render showOptionList = true - render component with "display: block" for dropdown list 1`] = `
<div
  className="multiple-select-container"
  id="multiple-select-24"
>
  <button
    className="multiple-select-default multiple-select-label"
    onClick={[Function]}
    type="button"
  >
    <span
      className="text-display"
      title="Select Options"
    >
      Select Options
    </span>
    <b
      className="caret"
    />
  </button>
  <div
    className="multiple-select-default multiple-select-options-container"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <ul
      className="multiple-select-options multiple-select-option-all"
    >
      <li
        className="multiple-select-item"
      >
        <input
          checked={true}
          className="option-checkbox"
          id="multiple-select-24-optionItemAll"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="option-label"
          htmlFor="multiple-select-24-optionItemAll"
        >
          (All)
        </label>
      </li>
    </ul>
    <ul
      className="multiple-select-options"
    >
      <span>
        No information
      </span>
    </ul>
  </div>
</div>
`;
